[3단계 - 데이터베이스 적용] 부산대 BE 김성문 미션 제출합니다.

doorcs opened the pull request:
2단계 코멘트를 보고, 스프링의 `Layered Architecture`에 대해 나름의 생각 정리를 해 봤습니다:

`Controller`: 데이터 입출력을 수행합니다.
- DTO 객체를 통해 사용자의 입력을 받음
- (Validation 라이브러리의 도움을 받아) 입력값 검증을 수행함
- 적절한 `Service` 계층의 함수를 호출함
- 서비스 계층에서 올려주는 값을 API 스펙에 맞는 형식으로 변환해 최종적으로 사용자에게 전달함

`Service`: 핵심 서비스 로직을 수행합니다.
- 컨트롤러로부터 받은 DTO 객체를 적절히 파싱해 도메인 객체를 만듦
- `Repository` 계층을 호출해 도메인 객체를 저장함
	- 수정의 경우, 조회 -> 수정 -> 저장
- `Repository` 계층을 호출해 데이터베이스로부터 도메인 객체를 조회함

`Repository`: 정보를 저장 및 조회합니다.
- 도메인 객체를 저장
- 도메인 객체를 조회

`컨트롤러`는 API단에서의 프론트엔드이니 입출력만을 담당하도록 하고, `리포지토리`는 DB(꼭 DBMS가 아니더라도)접근만 수행하도록 한다면
I/O, 데이터 저장 및 조회를 제외한 모든 작업들을 수행하고, `가장 복잡도가 높아질 수 있는 계층`이 `서비스`일까요?

여전히 제가 제대로 이해한 게 맞는지 확신이 없는데, 놓치고 있는 부분이나 더 고민해볼 수 있는 부분이 있다면 알려주시면 감사하겠습니다!

doorcs committed:
docs: 구현할 기능 목록 정리

doorcs committed:
refactor: 데이터베이스 마이그레이션

- HashMap -> H2 데이터베이스
- 정팩메 컨벤션에 따라 `.of()`, `.from()` 구분
- 객체 ID 부여 로직을 서비스가 담당하도록 수정
- 동시성 문제 해결을 위해 `AtomicLong` 도입
- 생성, 수정을 동시에 핸들링하기 위해 MERGE문 적용
- ReadProductResponse -> ProductResponse
- 상품 생성, 수정에 대한 예외 정의 및 핸들링

doorcs committed:
feat: SQL 스크립트 작성

jjongwa commented on src/main/java/gift/controller/GlobalExceptionHandler.java:
```diff
@@ -14,4 +16,14 @@ public class GlobalExceptionHandler {
         return ProblemDetail.forStatusAndDetail(HttpStatus.NOT_FOUND, e.getMessage());
     }
+
+    @ExceptionHandler(ProductCreateException.class)
+    public ProblemDetail handleProductCreateException(ProductCreateException e) {
+        return ProblemDetail.forStatusAndDetail(HttpStatus.INTERNAL_SERVER_ERROR, e.getMessage());
+    }
+
+    @ExceptionHandler(ProductUpdateException.class)
+    public ProblemDetail handleProductUpdateException(ProductUpdateException e) {
+        return ProblemDetail.forStatusAndDetail(HttpStatus.INTERNAL_SERVER_ERROR, e.getMessage());
+    }
```
오호.. ProblemDetail이란걸 사용해 주셨네요.!
어떤 장점이 있었나요? ProblemDetail을 사용하기 위해 추가해 주어야 설정 같은건 없을까요?

doorcs replied:
`ProblemDetail` 클래스의 정팩메를 통해 HTTP 에러코드와 메시지를 손쉽게 처리할 수 있어 좋았습니다.

레퍼런스 문서를 찾아보니 `spring.mvc.problemdetails.enabled=true` 옵션으로 `RFC 9457` Problem Detail을 적용할 수 있다고 하는데, 해당 옵션의 유무와 무관하게 `@RestControllerAdvice`에서 잡아주는 예외의 출력 형식은 동일한 것을 확인했습니다.

따라서 지금처럼 명시적으로 `ProblemDetail`을 리턴하는 경우 따로 추가해줘야 할 설정은 없으며, 해당 옵션은 `@Repository`가 DB 예외를 모두 `DataAccessException`으로 감싸주는 것과 유사하게 예외를 `ProblemDetail`로 감싸주는 역할을 하는 것으로 추측됩니다!

jjongwa commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -17,34 +20,47 @@ public class ProductService {
 
     private final ProductRepository productRepository;
 
+    private final static AtomicLong currentId = new AtomicLong(1);
```
AtomicLong을 사용해 동시성 이슈를 해결해 주셨네요 👍 

하지만 id 생성의 책임이 Service에 있는게 올바를까요? 
(지난 스텝부터 제 질문이 빙빙 돌리는 느낌이 들어 죄송하지만..ㅎㅎ)

이번 단게에서는 DB를 사용하게 되었는데요.! DB가 요런 문제의 해결법이 될 수도 있을  것 같네요~!

doorcs replied:
DB를 도임함으로써 DDL에 `AUTO_INCREMENT` 속성을 주고 `INSERT` 쿼리를 통해 DB에서 ID를 생성하도록 해줄 수 있었는데,
`static Long currentId`를 사용하던 코드를 리팩토링해야겠단 생각에 매몰되어 다른 접근방식을 떠올리지 못했던 것 같습니다.

ID는 DB가 부여하도록 하되, 엔티티 객체 생성에 대한 책임은 `Service`가 갖도록 하기 위해
`.of()` 정팩메를 오버로딩하여 `id`가 `null`인 객체를 생성할 수 있도록 수정했습니다!

jjongwa commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,8 @@
+    name     VARCHAR(100),
+    price    INT,
+    imageUrl VARCHAR(255),
+    PRIMARY KEY (id)
```
hint: `AUTO_INCREMENT`

doorcs replied:
바로 적용해줬습니다!

jjongwa commented on src/main/resources/data.sql:
기본 데이터까지 세팅해 주셨네요~! 👏

jjongwa commented on src/main/resources/application.properties:
```diff
@@ -1 +1,3 @@
 spring.application.name=spring-gift
+spring.datasource.url=jdbc:h2:mem:test
+spring.h2.console.enabled=true
```
spring.sql.init.mode는 필요 없을까요?

doorcs replied:
`spring.sql.init.mode`의 기본값은 `embedded`로,
H2와 같은 인메모리 DB를 사용할 경우에만 DB 초기화를 수행해줍니다.

지금은 요구사항에 따라 H2 DB를 사용하고 있어 기본값으로도 충분하지만,
DB 테이블을 초기화한다는 의도를 명확하게 드러내고 추후 변경에도 유연하게 대응할 수 있도록
`spring.sql.init.mode=always` 옵션을 추가해줬습니다!

jjongwa commented on src/main/java/gift/repository/ProductRepository.java:
```diff
@@ -1,52 +1,63 @@
 public class ProductRepository {
 
-    private final Map<Long, Product> products = new HashMap<>();
+    private final JdbcClient client;
```
다른 대안으로는 JdbcTemplate, SimpleJdbcInsert 등이 있었을 텐데요..!
그 중 JdbcClient를 선택한 이유는 무엇인가요?

doorcs replied:
`JdbcTemplate`은 카테캠 1단계 과정에서 사용해 본 경험이 있어서,
새로운 기술을 학습해보기 위해 이번에는 `JdbcClient`를 선택했습니다.

`VALUES (?, ?, ?)`대신 `VALUES (:name, :price, :imageUrl)`같은 named parameter를 사용할 수 있다는 점도 좋고,
메서드 체이닝 방식도 마음에 쏙 들어서 앞으로도 Jdbc를 다뤄야 할 일이 생긴다면 `JdbcClient`를 먼저 고려할 것 같습니다!

jjongwa commented on src/main/java/gift/repository/ProductRepository.java:
```diff
@@ -1,52 +1,63 @@
     public List<Product> findAll() {
-        return products.values()
+        var sql = "SELECT id, name, price, imageUrl FROM product";
+        return client.sql(sql)
+            .query(getMemberRowMapper())
             .stream()
```
이왕 JdbcClient를 쓴 김에 , 
```
return client.sql(sql)
             .query(getMemberRowMapper())
             .list();
```
로 바꿔봐도 되겠네요~!

doorcs replied:
`.query()`의 리턴타입인 `MappedQuerySpec`에는 `.toList()`를 사용할 수 없는것을 보고 스트림으로 변환했었는데,
`.toList()` 대신 `.list()`가 있었네요..

`JdbcClient` 인터페이스에 대해서도 추가적으로 공부해볼 수 있도록 하겠습니다!

jjongwa commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -17,34 +20,47 @@ public class ProductService {
+            request.imageUrl());
+        productRepository.save(updatedProduct);
+
+        return UpdateProductResponse.from(updatedProduct);
```
아하.. 지난 리뷰 이후로 이쪽이 이렇게 변경되었군요
지난번 코멘트가 오히려 혼동을 준 것 같네요..

이제 DB와 Jdbc를 사용하고 있으니 Repository 단에서 새로운 update 메서드를 만들고, Service 단에서도 find와 save를 호출하는 대신 update 하나만 써도 가능하게끔 바꿔 볼까요?

doorcs replied:
`findById()` 메서드 없이 `update()` 메서드만 단독으로 사용할 경우, 코드 구조가 깔끔해진다는 장점이 있지만 컨트롤러에서 `PathVariable`로 받아온 `id`가 유효하지 않은 경우에도 `ProductNotFoundException`를 발생시킬 수 없다는 문제가 있습니다.

`id`의 유효성을 검증하여 `DataAccessException` 대신 semantic한 예외를 발생시킬 수 있도록,
JPA를 참고하여 리포지토리에 `existsById()` 메서드를 추가하고 `수정 및 삭제` 메서드에 적용했습니다.

jjongwa reviewed (changes requested):
3단계도 잘 구현해 주셨어요~! 몇가지 간단한 코멘트 남겼습니다!

> 여전히 제가 제대로 이해한 게 맞는지 확신이 없는데, 놓치고 있는 부분이나 더 고민해볼 수 있는 부분이 있다면 알려주시면 감사하겠습니다!

이렇게까지 잘 정리해 주셨는데 아직도 확신이 없으면 안되죠~!ㅎㅎ 
다음번에도 요 이해한 내용을 기반으로 계층간 책임을 생각하면서 작성해 주시면 됩니다!

> 가장 복잡도가 높아질 수 있는 계층이 서비스일까요?

요 생각도 아주 바람직합니다! 
지금까지의 생각의 흐름으로 봤을 때 controller와 repository의 역할을 제외하면 나머진 비즈니스 로직은 다 서비스가 가져가기 때문에 가장 복잡하다고 생각되는게 당연한데요..! 

앞으로 조금씩 더 복잡한 조건의 미션을 마주하게 되었을 때, 복잡해 보일 수 있는 서비스 로직 안에서 각 도메인 객체가 책임을 가져갈 수 있는 부분이 있을지 고민해 보는 것도 좋은 공부가 될 것 같습니다!!

남긴 리뷰에 대한 답변 & 반영 후 다시 요청 주세요~!

doorcs committed:
chore: 객체의 id 부여를 데이터베이스가 책임지도록 수정

doorcs committed:
chore: 데이터베이스를 명시적으로 초기화하도록 수정

doorcs committed:
refactor: id가 null인 객체를 생성할 수 있도록 정팩메 오버로딩

doorcs committed:
refactor: 리포지토리 리팩토링

- 쿼리 가독성 향상을 위해 raw string 적용
- 엔티티 객체의 id를 데이터베이스가 부여하도록 쿼리 수정
- save 메서드가 데이터베이스에서 부여한 엔티티 id를 리턴하도록 수정
- update 메서드 구현
- 예외 처리를 위해 delete 메서드가 query affected row를 리턴하도록 수정

doorcs committed:
refactor: 서비스 리팩토링

- 엔티티 객체 ID 부여 책임을 데이터베이스가 갖도록 수정
- `수정 및 삭제` 메서드가 semantic한 예외를 발생시킬 수 있도록 수정
- `@Transactional` 어노테이션 추가

doorcs commented:
남겨주신 코멘트를 참고하며 공부하다 추가로 생긴 궁금증에 대해 질문 드리고 싶습니다:

- 이번 미션에서 제가 사용한 `ProblemDetail`과 `JdbcClient`의 경우 `스프링 6.x`, `스프링부트 3.x`부터 지원되기 시작한 기능인데
이렇게 최신 표준이나 신기술을 학습해야 할 경우, 멘토님께서는 어떤 접근 방식을 선호하시나요?

  - 일단 최신 표준만 제대로 익혀두고, 예전 기술은 나중에 필요하다면 배우도록 하자
  - 예전 기술로는 어떤식으로 구현했고, 어떤게 불편했는지부터 자세히 배운 다음에 최신 기술을 익히자
- DB를 사용할 경우 DB가 id를 부여하도록 하는 것이 가장 자연스럽다는 것은 이해가 되지만,
왜 `서비스`나 `리포지토리`에서 엔티티 객체의 id를 부여하는 것이 올바른 책임이 아닌지는 사실 아직도 잘 모르겠습니다..
DB 없이 엔티티 객체의 id를 직접 부여해줘야 하는 경우에도 id 생성 책임이 `서비스`나 `리포지토리`로부터 분리되어야 한다면,
서비스에서 사용할 `ProductIdGenerator`를 만들어 주는 것이 올바른 책임일까요?

---
멘토님 덕분에 `Layered Architecture`와 `올바른 책임`에 대해 깊게 고민해볼 수 있었습니다.
바쁘신 와중에도 매번 섬세한 리뷰 남겨주셔서 **정말 감사합니다!**

doorcs committed:
refactor: 관리자 페이지가 올바른 HTTP 메서드를 사용하도록 수정

jjongwa commented:
@doorcs 👍 

> 이번 미션에서 제가 사용한 ProblemDetail과 JdbcClient의 경우 스프링 6.x, 스프링부트 3.x부터 지원되기 시작한 기능인데
이렇게 최신 표준이나 신기술을 학습해야 할 경우, 멘토님께서는 어떤 접근 방식을 선호하시나요?

사실 요런 정도는 뭔가 배운다기 보다는 그냥 쓰다 보면 자연스럽게 익혀지게 되는 수준 같아요.! 그리고 `최신 표준이나 신기술을 학습해야 할 경우` 이러한 상황 자체가 결국 이전의 기술은 사용할 수 없는 상황이라 결국 최신 기술을 적용해야 한다고 볼 수도 있구요..!
-> 결론은 굳이 깊게 생각하지 말고 **필요하면 바로바로 써먹으면서 배운다** 가 될 것 같네요..!

> 왜 서비스나 리포지토리에서 엔티티 객체의 id를 부여하는 것이 올바른 책임이 아닌지는 사실 아직도 잘 모르겠습니다..

요것도 정말 간단하게 생각해 볼까요! 2단계까지는 DB라는게 없었기 때문에 Repository가 로컬 메모리 저장 & 조회 기능을 담담했고, 그래서 id 생성 역할이 Repository에 있어도 그렇게 어색하진 않습니다! (하지만 비즈니스 로직을 처리하는 서비스 단에선 어색하겠죠..! `물건을 저장한다`라는 기능이 굳이 `물건의 아이디는 겹치면 안되고 생성될 때 가장 최근에 만들어진 물건 id + 1에 해당한다`는 의미까지 알 필요 없는 것 처럼요)

말하신대로 ProductIdGenerator를 만들고 Repository가 ProductIdGenerator를 가지고 있는 것도 전 합당하다고 생각합니다~!

어떤 로직이 어느 계층의 책임인가도 결국 정답은 없고 각자의 의견 차가 있을 수 있어요! 같은 팀원들이 모두 동의한다면 그게 팀에서는 맞는 방향이 되는 겁니다~!

jjongwa reviewed (approved):
👍 잘 반영해 주셨어요~!
이번 미션은 여기서 종료하겠습니다!

이미 다음 미션 시작하셨을텐데, 여기서 배웠던 내용을 잊지 않고 다음 미션에서도 잘 적용해 주세요~!

고생 많으셨습니다!! 🙇
